### PR TITLE
Implemented subset plots

### LINKED ISSUE #23: Subset evaluation data for plots
### Instructions

Split cell-level metrics by channel type and cell type (not jointly) and use a facet plot to plot metrics for all subsets. Some useful visualizations: a separate precision/recall/f1 plot for each channel, aggregated over all cell types. A separate precision/recall/f1 plot for each cell type, aggregated over all channels. A separate precision/recall/f1 plot for each channel in each cell type. A separate precision/recall/f1 plot for each channel in each sample.

**Relevant background**

To further get insights about the model performance we want to more granularly subset the metrics.

**Design overview**

Take the `pred_list ` or an object of class `HDFLoader` and collapse all `activity_df`s into one big `activity_df`. Then a function that takes in the `activity_df` and a list of subset columns and returns plots of metrics for all subsets.

**Code mockup**

```
def collapse_activity_df(pred_list):
    out_df = pd.DataFrame()
    for sample in pred_list:
        out_df = pd.concat(out_df, sample["activity_df"])
    return out_df
```

```
def subset_plots(activity_df, subset_list, save_dir, save_file):
    ndim = len(subset_list)
    subset_uniques = {}
    for subset in subset_list:
        subset_uniques[subset] = np.unique(getattr(activity_df, subset))
    if ndim == 1:  
        plot_dim = [np.ceil(np.sqrt(len(subset_uniques[subset])))]*2
    if ndim>1:
        plot_dim = [len(item) for item in subset_uniques.items()]
    fig, ax = plt.subplots(**plot_dim, figsize=(plot_dim[0]*2, plot_dim[1]*2))
    for i in range(plot_dim[0]):
        for j in range(plot_dim[1]):
            metrics = calc_metrics(subset_activity_df(activity_df, subset_uniques, i, j)
            ax[i,j] = plot_metrics_against_threshold(metrics)
    plt.savefig()

```

```
def subset_activity_df(activity_df, subset_uniques, i, j):
    return activity_df_subset

```
**Required inputs**

`pred_iterator`: list or HDFLoader containing prediction dictionaries. Those should contain `activity_df` with columns `gt`, `prediction`, `cell_type`, `marker`, `dataset` or similar

`subset_list`: list that contains column names from the `activity_df` for which we want to subset the data and generate plots

**Output files**

Facet plots, for `len(subset_list)==1` the plots should be roughly squared so columns=rows=~sqrt(len(subset_list)) and for `len(subset_list)==2` the plots should be `columns = len(set(getattr(activity_df, subset_list[0])))` and `rows = len(set(getattr(activity_df, subset_list[0])))`

**Timeline**
Give a rough estimate for how long you think the project will take. In general, it's better to be too conservative rather than too optimistic. 
- [x] A couple days
- [ ] A week
- [ ] Multiple weeks. For large projects, make sure to agree on a plan that isn't just a single monster PR at the end. 

Estimated date when a fully implemented version will be ready for review: 10/17-19/2022

Estimated date when the finalized project will be merged in: 10/19/2022

### PR DESCRIPTION
**Pull request by JLrumberger**
**What is the purpose of this PR?**

This PR closes #23 . It contains code to prepare facet plots for arbitrary subsets of the cell-level predictions to get a clearer picture of error cases.

**How did you implement your changes**

I added functions `collapse_activity_dfs`, `subset_activity_df` and `subset_plots` to plot_utils.py and changed `calc_metrics` so that it does not internally defines function `calc_scores`, instead `calc_scores` is defined as a normal function so that I can get imported by `subset_plots`. 

**Remaining issues**

None

### CODE REVIEW

**Comment by JLrumberger** [comment]
Found a small error when applying the code on real data, so I'll upload an updated version within the next couple of minutes.

File: cell_classification/plot_utils.py [hunk 1]
```diff
@@ -188,6 +190,105 @@ def plot_metrics_against_threshold(
...
+    Args:
+        pred_list (list):
+            list of dictionaries containing activity dataframes
+    Returns:
+        collapsed_df (pd.DataFrame):
+            collapsed activity dataframe
+    """
+    collapsed_df = pd.DataFrame()
+    for sample in pred_list:
+        collapsed_df = pd.concat([collapsed_df, sample["activity_df"]])
```
Comment (ngreenwald) [inline cell_classification/plot_utils.py]: It's faster to make a list and then perform a single concatenate call at the end https://github.com/angelolab/ark-analysis/issues/751

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good, just a small tweak

**Review by ngreenwald — APPROVED** [review]
Looks good!

**Pull request merged**